pkg/kademlia/testrouting/testrouting.go

```diff
@@ -0,0 +1,308 @@
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// don't add ourselves
+	if node.Id == t.self {
+		return nil
+	}
+
+	// if the node is already here, update it
+	if cell, exists := t.nodes[node.Id]; exists {
+		cell.node = node
+		cell.lastUpdated = time.Now()
+		cell.fails = 0
+		// skip placement order and cache status
+		return nil
+	}
+
+	// add unconditionally (it might be going into a replacement cache)
+	t.nodes[node.Id] = &nodeData{
+		node:        node,
+		ordering:    t.counter,
+		lastUpdated: time.Now(),
+		fails:       0,
+
+		// makeTree within preserveInvariants might promote this to true
+		inCache: false,
+	}
+	t.counter++
+
+	t.preserveInvariants()
+	return nil
+}
+
+// ConnectionFailed should be called whenever a node can't be contacted.
+// If a node fails more than allowedFailures times, it will be removed from
+// the routing table. The failure count is reset every successful connection.
+func (t *Table) ConnectionFailed(node *pb.Node) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// if the node exists and the failure is with the address we have, record
+	// a failure
+
+	if data, exists := t.nodes[node.Id]; exists &&
+		pb.AddressEqual(data.node.Address, node.Address) {
+		data.fails++ //TODO: we may not need this
+		// if we've failed too many times, remove the node
+		if data.fails > t.allowedFailures {
+			delete(t.nodes, node.Id)
+
+			t.preserveInvariants()
+		}
+	}
+	return nil
+}
+
+// FindNear will return up to limit nodes in the routing table ordered by
+// kademlia xor distance from the given id.
+func (t *Table) FindNear(id storj.NodeID, limit int, restrictions ...pb.Restriction) ([]*pb.Node, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// find all non-cache nodes
+	nodes := make([]*nodeData, 0, len(t.nodes))
+	for _, node := range t.nodes {
+		if !node.inCache {
+			nodes = append(nodes, node)
+		}
+	}
+
+	// sort by distance
+	sort.Sort(nodeDataDistanceSorter{self: id, nodes: nodes})
+
+	// return up to limit nodes
+	if limit > len(nodes) {
+		limit = len(nodes)
+	}
+	rv := make([]*pb.Node, 0, limit)
+	for _, data := range nodes[:limit] {
+		rv = append(rv, data.node)
+	}
+	return rv, nil
+}
+
+// Local implements the dht.RoutingTable interface
+func (t *Table) Local() pb.Node {
+	// the routing table has no idea what the right address of ourself is,
+	// so this is the wrong place to get this information. we could return
+	// our own id only?
+	panic("Unimplementable")
+}
+
+// Self returns the node's configured node id.
+func (t *Table) Self() storj.NodeID { return t.self }
+
+// MaxBucketDepth returns the largest depth of the routing table tree. This
+// is useful for determining which buckets should be refreshed.
+func (t *Table) MaxBucketDepth() (int, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	var maxDepth int
+	t.walkLeaves(t.makeTree(), func(b *bucket) {
+		if b.depth > maxDepth {
+			maxDepth = b.depth
+		}
+	})
+	return maxDepth, nil
+}
+
+// GetNodes implements the dht.RoutingTable interface
```

Comment:
X implements Y doesn't explain much in documentation. That could be an extra comment, but the documentation should describe the thing it does.